lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp

```diff
@@ -173,6 +173,72 @@ struct StorageGetOpLowering : public OpConversionPattern<arc::StorageGetOp> {
   }
 };
 
+struct AssertOpLowering : public OpConversionPattern<arc::AssertOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::AssertOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+
+    Value negateCondition = rewriter.create<arith::XOrIOp>(
+        op.getLoc(), adaptor.getProperty(),
+        rewriter.create<arith::ConstantOp>(op.getLoc(),
+                                           rewriter.getBoolAttr(true)));
+
+    rewriter.replaceOpWithNewOp<scf::IfOp>(
+        op, negateCondition, [&](OpBuilder &builder, Location loc) {
+          // If the condition doesn't hold then print the message and abort
+          auto message = op.getMsg().value_or("");
+          auto module = op->getParentOfType<ModuleOp>();
+
+          LLVM::LLVMFuncOp putsFunc;
+          LLVM::GlobalOp globalOp;
+
+          // If there's no message just abort
+          if (!message.empty()) {
+            {
+              OpBuilder::InsertionGuard guard(rewriter);
+              rewriter.setInsertionPointToStart(module.getBody());
+              putsFunc = lookupOrCreateFn(
```

Comment:
While looking into what to suggest, looks like there are various useful helpers for basically exactly this flow.

Additionally, `cf.assert` has an LLVM lowering -- was wondering if makes sense to lower to/use `cf.assert` since this seems rather similar?

Regardless, the lowering there uses the rewriter-aware and useful helpers for grabbing a print-like method as well as global strings:

https://github.com/llvm/llvm-project/blob/9d4f7f44b64d87d1068859906f43b7ce03a7388b/mlir/lib/Conversion/ControlFlowToLLVM/ControlFlowToLLVM.cpp#L52

It doesn't re-use the global string definition, FWIW, but anyway interesting related bits if not already on your radar.

And actually, looks like it ultimately just calls `lookupOrCreateFn` (or wrappers) anyway: https://github.com/llvm/llvm-project/blob/cc54129b983799e1aaea77aa0ff3040dc30cbc8c/mlir/lib/Conversion/LLVMCommon/PrintCallHelper.cpp#L63 .  Not sure how creating a global is safe?

Seems like could do what this is doing using rewriter (instead of the builder it creates)?
https://github.com/llvm/llvm-project/blob/cc54129b983799e1aaea77aa0ff3040dc30cbc8c/mlir/lib/Dialect/LLVMIR/IR/FunctionCallUtils.cpp#L51
(possible upstream patch, but if nothing else locally if looking to address that...)